pkg/external_repos/introspect.go

```diff
@@ -132,3 +150,47 @@ func httpClient(useCert bool) (http.Client, error) {
 		return http.Client{}, nil
 	}
 }
+
+// UpdateIntrospectionStatusMetadata updates introspection timestamps, error, and status on repo. Use after calling Introspect().
+func UpdateIntrospectionStatusMetadata(repo dao.Repository, repoDao dao.RepositoryDao, count int64, err error) error {
+	introspectTimeEnd := time.Now()
+	repo = updateIntrospectionStatusMetadata(repo, count, err, &introspectTimeEnd)
+
+	if err := repoDao.Update(repo); err != nil {
+		return fmt.Errorf("failed to update introspection timestamps: %w", err)
+	}
+
+	return nil
+}
+
+func updateIntrospectionStatusMetadata(input dao.Repository, count int64, err error, introspectTimeEnd *time.Time) dao.Repository {
+	output := input
+	output.LastIntrospectionTime = introspectTimeEnd
+
+	// If introspection was successful
+	if err == nil {
+		if count != 0 {
+			output.LastIntrospectionUpdateTime = introspectTimeEnd
+		}
+		output.LastIntrospectionSuccessTime = introspectTimeEnd
+		output.LastIntrospectionError = pointy.String("")
+		output.Status = dao.StatusValid
+		return output
+	}
+
+	// If introspection fails
+	switch input.Status {
+	case dao.StatusValid:
+		output.Status = dao.StatusUnavailable
```

Comment:
this line might be worth a comment, because i can't figure out why it makes sense :)

Reply:
if the status of the repository was valid before it failed, then the status becomes unavailable